### ISSUE TITLE
Difference in outputs between Lit-GPT and HuggingFace models. 

### ISSUE THREAD

**Issue by Andrei-Aksionov** [Lightning-AI/lit-gpt#914]
Hi there 👋 

> [!TIP]
>TLDR; Lit-GPT model produces a different output in comparison to HuggingFace model, despite having identical weights, because Lit-GPT uses `scaled_dot_product_attention` from PyTorch (which also implements FlashAttention), while HuggingFace uses either FlashAttention from the `flash-attn` package or, if it's not installed, theirs own implementation of attention scores calculation.

Lit model can produce a different output in comparison to HF model (with identical weights). 
For that case in tests we even have a [`pytest.xfail`](https://github.com/Lightning-AI/lit-gpt/blob/cca2986426988ea855112a78ed061228a13fb1d4/tests/test_model.py#L36) if the model isn't in `torch.float32` precision.
Some users have also noticed a difference: https://github.com/Lightning-AI/lit-gpt/issues/844#issuecomment-1868993712, #848. 

Since I'm working on a tasks that requires fully matching forward pass, I decided to dive deeply and, if possible, find a solution.
(Spoiler alert: I haven't found a solution to the problem 🙁).

For the debug I used Pythia model, that means that on the HF side the code is located [here](https://github.com/huggingface/transformers/blob/main/src/transformers/models/gpt_neox/modeling_gpt_neox.py). Also it looks like my findings can be applied to all other HF models.

So, after a really thorough comparison of each stage of forward pass for both Lit and HF models, I noticed that the difference is happening during calculation of attention scores. All tensors up to this point are identical between Lit and HF models.

The issue is that HF uses either [custom implementation](https://github.com/huggingface/transformers/blob/main/src/transformers/models/gpt_neox/modeling_gpt_neox.py#L250) or [FlashAttention](https://github.com/huggingface/transformers/blob/3d2900e829ab16757632f9dde891f1947cfc4be0/src/transformers/models/gpt_neox/modeling_gpt_neox.py#L425) (if it's installed). While Lit-GPT uses only [torch.nn.functional.scaled_dot_product_attention](https://github.com/Lightning-AI/lit-gpt/blob/cca2986426988ea855112a78ed061228a13fb1d4/lit_gpt/model.py#L225).
And depending on the `device`, `input data` and `precision` there might be a difference between HF attention and `scaled_dot_product_attention`. All the numbers can be found in this [notebook](https://github.com/Andrei-Aksionov/lit-gpt/blob/644b14beabd20e94c5d4f4ff27090879a2cd227a/notebooks/forward_attn_reproducible.ipynb).

In the notebook I compared outputs from HF attention, `scaled_dot_product_attention` and it's example implementation that one can find in [docs](https://pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html). In addition, I provided comparison numbers to flash attention implementation from `flash-attn` package (it's used in HF if it's installed).

From the notebook one can see that we can get identical results (between HF and `scaled_dot_product_attention`) in `torch.float32` and Uniform Distribution, both on CPU and GPU. With the same precision but a Normal Distribution (which all tensors have) we can see a divergence there. Though it's very small, which can explain why with this precision we can get similar outputs in tests.

With a lower precision (e.g., bfloat16/float16) the difference in the outputs is more noticeable, especially on a GPU - only 48% of tensor values are "close".

And such a difference is not only between HF's attn and `scaled_dot_product_attention`, but also between the code from the [docs](https://pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html), despite it says "# Efficient implementation equivalent to the following: ".

The biggest difference is between `scaled_dot_product_attention` and FlashAttention from `flash-attn` package: only 0.86% are "close" in `torch.float16` and on `cuda`.

----

Bottom line: since, in my opinion, we should continue to stick to `scaled_dot_product_attention`, there is no way we can make an output from Lit-GPT to be identical to HF. The best we can do is to make it very-very close in `torch.float32` precision.

----

Resources:
- [notebook](https://github.com/Andrei-Aksionov/lit-gpt/blob/644b14beabd20e94c5d4f4ff27090879a2cd227a/notebooks/forward_attn_reproducible.ipynb) with comparisons.

**Comment by Andrei-Aksionov** [comment]
I created an Issue in PyTorch repo: https://github.com/pytorch/pytorch/issues/119188
If it's indeed an unexpected behavior and will be fixed, then it will only make the output from HF's and LIt's attention similar in `torch.float32`, in other precisions there still be a difference anyway.
